qiskit/circuit/instruction.py

```diff
@@ -307,7 +307,49 @@ def broadcast_arguments(self, qargs, cargs):
             raise QiskitError(
                 'The amount of qubit arguments does not match the instruction expectation.')
 
+        if len(cargs) != self.num_clbits:
+            raise QiskitError(
+                'The amount of clbit arguments does not match the instruction expectation.')
+
         #  [[q[0], q[1]], [c[0], c[1]]] -> [q[0], c[0]], [q[1], c[1]]
         flat_qargs = [qarg for sublist in qargs for qarg in sublist]
         flat_cargs = [carg for sublist in cargs for carg in sublist]
         yield flat_qargs, flat_cargs
+
+    def power(self, exponent):
+        """
+        Creates an instruction with the gate repeated `exponent` amount of times.
+        Negative exponent creates an instruction with Gate.inverse().
+        Zero exponent creates an instruction with u3(0,0,0).
+
+        Args:
+            exponent (int): Gate^exponent
+
+        Returns:
+            Instruction: Containing the definition.
+
+        Raises:
+            QiskitError: If called by a non-integer exponent.
+        """
+        if int(exponent) != exponent:
+            raise QiskitError("Instruction only supports integer exponentiation.")
+        exponent = int(exponent)
+
+        instruction = Instruction(name="%s^%s" % (self.name, exponent),
```

Comment:
`type(self)(...)` where `self` is a `CnotGate` would return `CnotGate`. The most generic type we have is `Instruction` and we use it for custom composite gate, right?

Reply:
Good point, `type(self)(...)` is not what we want.

It would be nice though to be keep the return as a `Gate`, since they're assumed to be unitary. Otherwise, code like `op.power(x).to_matrix()` would be invalid, even if `op` is a `Gate`.

Reply:
Once we get composite gates as gates, we can do this https://github.com/Qiskit/qiskit-terra/issues/2896

Reply:
Solved in 925cc51d